### ISSUE TITLE
Control instantiation parameters are not always treated as constants

### ISSUE THREAD

**Issue by vgurevich** [p4lang/p4c#2995]
Here is a simple example:

```
control c()(int bar=16) {
    bit<(bar)> foo;
    apply {
    }
}
```

Compiling this with `p4test` produces the following error:
```
param_test.p4(2): [--Werror=expected] error: bar: expected a constant
    bit<(bar)> foo;
         ^^^
```

**Comment by jnfoster** [comment]
I think this is consistent with the current spec. It says:

> The following are compile-time known values:
> 
> * Integer literals, Boolean literals, and string literals.
> * Identifiers declared in an error, enum, or match_kind declaration.
> * The default identifier.
> * The size field of a value with type header stack.
> * The _ identifier when used as a select expression label
> * Identifiers that represent declared types, actions, tables, parsers, controls, or packages.
> * List expression where all components are compile-time known values.
> * Structure initializer expressions, where all fields are compile-time known values.
> * Instances constructed by instance declarations (Section 10.3) and constructor invocations.
> * The following expressions (+, -, *, / , %, cast, !, &, |, &&, ||, << , >> , ~ ,  >, <, ==, !=, <=, >=, ++, [:]) when their operands are all compile-time known values.
> * Identifiers declared as constants using the const keyword.
> * Expressions of the form e.minSizeInBits() and e.minSizeInBytes().

I don't see constructor parameters here. 

Maybe we should fix that, but then it's not a `p4c` bug, it's a spec bug.

**Comment by jnfoster** [comment]
I agree they are morally compile-time known.

But, to be a language spec lawyer: 
* We say the expressions filled in for constructor parameters must be compile-time known values.
* But we do not say that constructor parameters are themselves compile-time known values.
So technically `p4c` is doing the right thing (given the current specification).

**Comment by vgurevich** [comment]
@jnfoster -- in this case passing parameters from outer controls into the inner controls (for examples) is not allowed:

```
control inner()(int param = 10) {
    apply {
    }
}

control outer() (int param = 20) {
    inner(param) i;
    apply {
       i.apply();
    }
}
```

Should I file the bug that `inner()` is instantiated with the value of the parameter that is not known at compile-time? :)

**Comment by vgurevich** [comment]
@jnfoster -- also the spec says that table sizes must be compile-time only values. Should I file a bug against the compiler that currently allows this:

```
control c()(int param = 100) {
    action a() {}
    table t {
        size = param;
        actions = { a; }
    }
    apply {
        t.apply();
    }
}
```

**Comment by jnfoster** [comment]
Let me try to explain.

P4 has a set of expressions. There are infinitely many of them. They are the set of abstract syntax trees generated by the grammar we given in the appendix of the language specification. Let's call this set EXPR.

Of this infinite set of expressions, a subset are compile-time known. This set is the smallest one "closed" under the list I gave above. (So, technically, it's what's called an inductively-defined set). Let's call this set CTKV.

To typecheck a control instantiation, we insist that the constructor parameters be from the set CTKV. 

Example:
```
control c()(bit<8> x) { apply { ... x ... } }
c(1 + 2) c1;
```
Here, the expression `1+2` is an EXPR that is also in CTKV. But the parameter `x` is an EXPR that is *not* in CTKV. 

To put it another way, we insist that the *arguments* (like `1+2`) be CTVK but we don't say that the *parameter* itself is CTKV. 

Mathematically, this is a perfectly sensible thing to do. Practically, of course, it's ludicrous. 

So the simple fix is to add constructor parameters to the set CTKV.

**Comment by jnfoster** [comment]
Regarding this comment, https://github.com/p4lang/p4c/issues/2995#issuecomment-997288806, yes you should file a big.

The compiler is not checking that the size property is compile-time known at all.

See here: https://github.com/p4lang/p4c/blob/6b635f06e4b39a8a9d90a8bdd685bd97cd3baf6f/midend/checkSize.h#L25

The compiler front-end and mid-end also accept this program, which is wrong.

```
control c(bit<8> x)(int param = 100) {
    action a() {}
    action NoAction() {}
    table t {
        size = x;
        actions = { a; }
    }
    apply {
        t.apply();
    }
}
```

Probably real backends barf on it. But we should be rejecting the program earlier.

**Comment by vgurevich** [comment]
OK, so practically speaking, I think we need to fix this specific issue by assuming that https://github.com/p4lang/p4-spec/issues/1001 is a done deal. Are we in agreement? 

Moreover, the vast majority of the programs that use control instantiation parameters, use them to control the sizes of tables and externs. Thus, if we add the check before fixing this issue, pretty much all the code out there will be broken.

**Comment by mbudiu-vmw** [comment]
In principle it would be nice for this to work, but given the current structure of the compiler it's not easy. The problem is one of pass ordering: we do first constant folding, then type-checking, then type specialization for instantiations, and we never repeat. To propagate these constants you would have to put this in a loop and to modify the type-checking algorithm to allow for "constants whose values are not yet known", like these constructor parameters.

What I am saying is that the first time we typecheck the instantiation of the inner control we don't know the value of the constructur parameter.

**Comment by jnfoster** [comment]
I think we should decide if we want these to behave like macros (where any valid substitution is allowed) or like types (where the type checking is done statically and modularly). If the latter, then it may be best to keep the current design...

**Comment by vgurevich** [comment]
I would suggest making the macro-like functionality working as best as we can, since this is what is being currently used in practice (and quite widely), while leaving modularity concerns for the future, since it is not something that is being practically used/implemented and, frankly, will require tons of other things in the language that we do not have. 

So, type checking will be done statically, but not modularly.

**Comment by mbudiu-vmw** [comment]
This is a duplicate of #2025
